### PR TITLE
[ZEPPELIN-2538] JDBC completer improvements for work with large meta

### PR DESCRIPTION
**Pull request by tinkoff-dwh**
### What is this PR for?
There are some problems if meta is large (few schemas, each schema contains 500+ tables etc.).
Problems:

1. loading is very long
2. each update takes one connection if updates are long, the situation may arise that the entire pool will be busy
3. no cache
This PR solves these problems. Added cache and access by full path (schema.table, schema.table.column) + protection for release the connections

### What type of PR is it?
Improvement 

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-2538

### Screenshots (if appropriate)
peek 2017-05-15 15-03

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 1]
```diff
@@ -188,11 +195,43 @@ private void setMaxLineResults() {
...
+    } catch (InterruptedException e) {
+      logger.warn("Completion timeout", e);
+      if (connection != null) {
+        try {
+          connection.close();
+        } catch (SQLException e1) {
+          logger.warn("Error close connection", e1);
+        }
+      }
+    }
```
Comment (1ambda) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: Is it safe not to close the connection when InterruptedException is not thrown?

Comment (tinkoff-dwh) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: @1ambda Yes, if everything is OK the connection will close into completer.createOrUpdateFromConnection.

Comment (1ambda) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: Thanks.

https://github.com/tinkoff-dwh/zeppelin/blob/e770d261cd492dda4167f0c0b5c69916fb744109/jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java#L298

**Comment by Leemoonsoo** [comment]
\cc @cloverhearts @astroshim

**Comment by cloverhearts** [comment]
I will test on this pr.
Thank you for greate feature. :)

**Comment by astroshim** [comment]
I think this feature is really cool.
BTW I got following exception whenever i try to run sql completion.

```
==> logs/zeppelin-interpreter-jdbc-jdbc.log <==
ERROR [2017-05-25 01:29:57,441] ({pool-1-thread-5} TThreadPoolServer.java[run]:296) - Error occurred during processing of message.
java.lang.NullPointerException
	at org.apache.zeppelin.jdbc.SqlCompleter.completeKeyword(SqlCompleter.java:407)
	at org.apache.zeppelin.jdbc.SqlCompleter.completeName(SqlCompleter.java:469)
	at org.apache.zeppelin.jdbc.SqlCompleter.complete(SqlCompleter.java:105)
	at org.apache.zeppelin.jdbc.JDBCInterpreter.completion(JDBCInterpreter.java:844)
	at org.apache.zeppelin.interpreter.LazyOpenInterpreter.completion(LazyOpenInterpreter.java:136)
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreterServer.completion(RemoteInterpreterServer.java:573)
	at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Processor$completion.getResult(RemoteInterpreterService.java:1940)
	at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Processor$completion.getResult(RemoteInterpreterService.java:1925)
	at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)
	at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
	at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:285)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

==> logs/zeppelin-.local.log <==
ERROR [2017-05-25 01:29:57,443] ({qtp517210187-14} NotebookServer.java[onMessage]:383) - Can't handle message
org.apache.zeppelin.interpreter.InterpreterException: org.apache.thrift.transport.TTransportException
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.completion(RemoteInterpreter.java:493)
	at org.apache.zeppelin.interpreter.LazyOpenInterpreter.completion(LazyOpenInterpreter.java:136)
	at org.apache.zeppelin.notebook.Paragraph.completion(Paragraph.java:304)
	at org.apache.zeppelin.notebook.Note.completion(Note.java:664)
	at org.apache.zeppelin.socket.NotebookServer.completion(NotebookServer.java:1345)
	at org.apache.zeppelin.socket.NotebookServer.onMessage(NotebookServer.java:330)
	at org.apache.zeppelin.socket.NotebookSocket.onWebSocketText(NotebookSocket.java:59)
	at org.eclipse.jetty.websocket.common.events.JettyListenerEventDriver.onTextMessage(JettyListenerEventDriver.java:128)
	at org.eclipse.jetty.websocket.common.message.SimpleTextMessage.messageComplete(SimpleTextMessage.java:69)
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.appendMessage(AbstractEventDriver.java:65)
	at org.eclipse.jetty.websocket.common.events.JettyListenerEventDriver.onTextFrame(JettyListenerEventDriver.java:122)
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.incomingFrame(AbstractEventDriver.java:161)
	at org.eclipse.jetty.websocket.common.WebSocketSession.incomingFrame(WebSocketSession.java:309)
	at org.eclipse.jetty.websocket.common.extensions.ExtensionStack.incomingFrame(ExtensionStack.java:214)
	at org.eclipse.jetty.websocket.common.Parser.notifyFrame(Parser.java:220)
	at org.eclipse.jetty.websocket.common.Parser.parse(Parser.java:258)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.readParse(AbstractWebSocketConnection.java:632)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:480)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.thrift.transport.TTransportException
	at org.apache.thrift.transport.TIOStreamTransport.read(TIOStreamTransport.java:132)
	at org.apache.thrift.transport.TTransport.readAll(TTransport.java:86)
	at org.apache.thrift.protocol.TBinaryProtocol.readAll(TBinaryProtocol.java:429)
	at org.apache.thrift.protocol.TBinaryProtocol.readI32(TBinaryProtocol.java:318)
	at org.apache.thrift.protocol.TBinaryProtocol.readMessageBegin(TBinaryProtocol.java:219)
	at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:69)
	at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Client.recv_completion(RemoteInterpreterService.java:372)
	at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Client.completion(RemoteInterpreterService.java:355)
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.completion(RemoteInterpreter.java:488)
	... 21 more
```

Could you take a look at this?

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java [hunk 2]
```diff
@@ -422,7 +401,7 @@ public void initFromConnection(Connection connection, String schemaFiltersString
    * @return -1 in case of no candidates found, 0 otherwise
    */
   private int completeKeyword(String buffer, int cursor, List<CharSequence> candidates) {
-    return keywordCompleter.complete(buffer, cursor, candidates);
+    return keywordCompleter.getCompleter().complete(buffer, cursor, candidates);
   }
 
   /**
```
Comment (tinkoff-dwh) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: for link

**Comment by tinkoff-dwh** [comment]
@astroshim 
it is strangely. do you package new jar? 407 line SqlConleter.java it is start of comment (multiline)

**Comment by astroshim** [comment]
@tinkoff-dwh Sorry it was my problem.
Tested and working good!

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by astroshim** [comment]
@cloverhearts Did you test this PR?

**Comment by cloverhearts** [comment]
@tinkoff-dwh 
Sorry. My confirmation is late.
I will left comment for results in until 12 hours.
I'm sorry for wait.

**Comment by cloverhearts** [comment]
Tested.
Well working :)
Thanks great feature.

exists some bug on web. (can not found moment and autocompletion `_` problem.)
but, not related this pr.

LGTM

**Comment by cloverhearts** [comment]
If there is no more comment on this, I will merge it.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 3]
```diff
@@ -188,11 +195,43 @@ private void setMaxLineResults() {
...
+    int ttlInSeconds = Integer.valueOf(
+        StringUtils.defaultIfEmpty(getProperty(sqlCompleterTtlKey), DEFAULT_COMPLETER_TTL)
+    );
+    final SqlCompleter completer;
+    if (sqlCompleter == null) {
+      completer = new SqlCompleter(ttlInSeconds);
+    } else {
+      completer = sqlCompleter;
+    }
+    ExecutorService executorService = Executors.newFixedThreadPool(1);
```
Comment (jongyoul) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: Not sure if it's good to make a new thread pool whenever called this method.

Comment (tinkoff-dwh) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: @jongyoul 
it is protection for connections pool. This pool terminated by timeout (if process of generate completions is long).

Comment (jongyoul) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: I know what you intend. But I just wonder the way to reuse the threadpool again. What do you think of it?

Comment (masyan) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: if shared the interpreter then there will be problems. `awaitTermination` interrupt all tasks

Comment (tinkoff-dwh) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: if shared the interpreter then there will be problems. awaitTermination interrupt all tasks

Comment (tinkoff-dwh) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: @jongyoul 
if I answered your question, could you re-initiate merge?

**Comment by jongyoul** [comment]
LGTM. Will merge it now.

**Pull request closed without merging**